karapace/schema_reader.py

```diff
@@ -285,9 +300,77 @@ def handle_messages(self) -> None:
                     self.offset_watcher.offset_seen(self.offset)
 
+            self._report_schema_metrics(
+                schema_records_processed_keymode_canonical,
+                schema_records_processed_keymode_deprecated_karapace,
+            )
+
+    def _report_schema_metrics(
+        self,
+        schema_records_processed_keymode_canonical: int,
+        schema_records_processed_keymode_deprecated_karapace: int,
+    ) -> None:
+        # Update processing counter always.
+        self.stats.increase(
+            metric=METRIC_SCHEMA_TOPIC_RECORDS_PROCESSED_COUNT,
+            inc_value=schema_records_processed_keymode_canonical,
+            tags={"keymode": KeyMode.CANONICAL},
+        )
+        self.stats.increase(
+            metric=METRIC_SCHEMA_TOPIC_RECORDS_PROCESSED_COUNT,
+            inc_value=schema_records_processed_keymode_deprecated_karapace,
+            tags={"keymode": KeyMode.DEPRECATED_KARAPACE},
+        )
+
+        # Update following gauges only if there is a possibility of a change.
+        records_processed = bool(
+            schema_records_processed_keymode_canonical or schema_records_processed_keymode_deprecated_karapace
+        )
+        if records_processed:
+            self.processed_canonical_keys_total += schema_records_processed_keymode_canonical
+            self.stats.gauge(
+                metric=METRIC_SCHEMA_TOPIC_RECORDS_PER_KEYMODE_GAUGE,
+                value=self.processed_canonical_keys_total,
+                tags={"keymode": KeyMode.CANONICAL},
+            )
+            self.processed_deprecated_karapace_keys_total += schema_records_processed_keymode_deprecated_karapace
+            self.stats.gauge(
+                metric=METRIC_SCHEMA_TOPIC_RECORDS_PER_KEYMODE_GAUGE,
+                value=self.processed_deprecated_karapace_keys_total,
+                tags={"keymode": KeyMode.DEPRECATED_KARAPACE},
+            )
+
+            self.stats.gauge(metric=METRIC_SCHEMAS_GAUGE, value=len(self.schemas))
+            self.stats.gauge(metric=METRIC_SUBJECTS_GAUGE, value=len(self.subjects))
+            for _, subject_data in self.subjects.items():
+                schema_data = subject_data.get("schemas", {})
+                versions = 0
+                soft_deleted_versions = 0
+                for version in schema_data.values():
+                    if not version.get("deleted", False):
+                        versions += 1
+                    else:
+                        soft_deleted_versions += 1
+                self.stats.gauge(
+                    metric=METRIC_SUBJECT_DATA_SCHEMA_VERSIONS_GAUGE,
+                    value=versions,
+                    tags={"state": "live"},
+                )
+                self.stats.gauge(
+                    metric=METRIC_SUBJECT_DATA_SCHEMA_VERSIONS_GAUGE,
+                    value=soft_deleted_versions,
+                    tags={"state": "soft_deleted"},
+                )
```

Comment:
After removal of `subject` from tags, these calls should be called once, not per `for ... self.subjects.item()` loop iteration, right? Otherwise value overrides other subject's value.